invenio_app_rdm/upgrade_scripts/migrate_4_0_to_6_0.py

```diff
@@ -0,0 +1,350 @@
+    schemes = VocabularyScheme.query.filter_by(parent_id="affiliations").all()
+    for scheme in schemes:
+        db.session.delete(scheme)
+    db.session.delete(VocabularyType.query.filter_by(id="affiliations").one())
+    db.session.commit()
+
+
+def check_affiliations():
+    """Checks the migration readiness of creatibutor affiliations."""
+
+    total = 0
+    def _should_be_vocabulary(creatibutors):
+        """Checks the schema of the affiliation.
+
+        If an affiliation has identifiers but no ROR, it is not valid.
+        """
+        vocab_affs = []
+        for creatibutor in creatibutors:
+            affiliations = creatibutor.get("affiliations", [])
+            for aff in affiliations:
+                ids = aff.get("identifiers", [])
+                ror = None
+                for id_ in ids:
+                    if id_.get("scheme") == "ror":
+                        ror = id_["identifier"]
+                        break
+                if not ror and ids:
+                    total += 1
+                    vocab_affs.append(aff["name"])
+
+        return vocab_affs
+
+    print("Checking for affiliations migration readiness...")
+    invalid_affiliations_rec = {}
+    for record_metadata in RDMRecord.model_cls.query.all():
+        record = RDMRecord(record_metadata.data, model=record_metadata)
+        # publish record subjects take presedence if id is repeated
+        # | operator is only available from py 3.9 on
+        inv_affs_creators = _should_be_vocabulary(
+            record.get("metadata").get("creators", [])
+        )
+        inv_affs_contributors = _should_be_vocabulary(
+            record.get("metadata").get("contributors", [])
+        )
+
+        if inv_affs_creators or inv_affs_contributors:
+            invalid_affiliations_rec[record["id"]] = {
+                "creators": inv_affs_creators,
+                "contributors": inv_affs_contributors
+            }
+
+    invalid_affiliations_draft ={}
+    for draft_metadata in RDMDraft.model_cls.query.all():
+        # Skipping deleted drafts because can't be committed
+        if draft_metadata.is_deleted:
+            continue
+
+        draft = RDMDraft(draft_metadata.data, model=draft_metadata)
+        inv_affs_creators = _should_be_vocabulary(
+            draft.get("metadata").get("creators", [])
+        )
+        inv_affs_contributors = _should_be_vocabulary(
+            draft.get("metadata").get("contributors", [])
+        )
+        if inv_affs_creators or inv_affs_contributors:
+            invalid_affiliations_draft[draft["id"]] = {
+                "type": "draft",
+                "creators": inv_affs_creators,
+                "contributors": inv_affs_contributors
+            }
+
+    invalid_affiliations = {}
+    if invalid_affiliations_rec:
+        invalid_affiliations["record"] = invalid_affiliations_rec
+    if invalid_affiliations_draft:
+        invalid_affiliations["draft"] = invalid_affiliations_draft
+
+    print(
```

Comment:
Let's refactor this part a bit to something like
```python
if invalid_affiliations:
        print(
            f"Your instance has {total} affiliations that need to be fixed. " +
            "Check the invalid_affiliation.yaml file for more details."
        )
        with open('invalid_affiliations.yaml', 'w') as f:
            yaml.dump(list(invalid_affiliations), f)
else:
        print(f"Your instance has no affiliations that need to be fixed.")
```

Reply:
We should also keep a `ROR` counter to tell the user if they need to index the ROR vocabulary or they can do with only free text affiliations.

Reply:
Done :)